### DISCUSSION TITLE
Different Docker Version in Windows and Rancher Desktop VM

### DISCUSSION THREAD

**Discussion by ISimionEndava** [rancher-sandbox/rancher-desktop#5856] [Q&A] [answered]
Hi, I installed Rancher Desktop (no prior Docker install/Docker Desktop install or anything like that) on **Windows 11**. I installed it with the **dockerd** option, I checked the docker version on Powershell, and I got **24.0.6**; I entered the Rancher Desktop VM (_rdctl shell_) and checked the docker version, and I got **23.0.6**; for docker-compose, I got the same version 2.21.0 both on Powershell and on Rancher Desktop VM (_rdctl shell_). 

1. How come I have a different docker version? I would expect the docker version to be the same, wouldn't I?
2. Can I update the docker version provided by Rancher Desktop to the latest version (24.0.7) and docker-compose to the latest version (2.23.0)? And if yes, how can I achieve that?

**Comment by mook-as** [comment, marked as answer]
There's two parts, the docker client and the docker server.  Currently we're using docker client 24.0.6 but docker server 23.0.6 (because the latter is using alpine packages).  And `docker-compose` is a separate thing altogether.

As for upgrading:
- I don't think we have anything that we can support.  You can try things, but if things break you'll have to figure out how to fix them yourself.
- Upgrading the server is unlikely to be possible; Alpine 3.18 only has [docker-engine 23.0.6](https://pkgs.alpinelinux.org/packages?name=docker-engine&branch=v3.18). You may be able to force install the version from Alpine edge, but that might break the install.
- For the clients (`docker`, `docker-compose`) you can try getting binaries from elsewhere and using it instead of the ones shipped by Rancher Desktop.  For example, you may be able to get things from [docker-compose releases](https://github.com/docker/compose/releases).

Again, if things break you'll have to figure out the issue yourself, especially if you change the dockerd version manually.

**Reply by ISimionEndava** [reply]
Thank you for getting back to me so quickly; the answers are helpful. 

1. You are saying `docker-compose` is a separate thing altogether; can you expand more on that, please?
2. If I am successful in updating `docker` and `docker-compose` for both the client and the server, what will happen when a new Rancher Desktop update gets released? I expect to get the versions for both client and server from the new Rancher Desktop update, or would I still stay with the versions I updated manually?

**Reply by mook-as** [reply]
1. `docker-compose` is a separate binary that is built from https://github.com/docker/compose — the `docker` CLI runs it as a plugin. It's placed in `%USERPROFILE%\.docker\cli-plugins` and the `docker` CLI will find it based on it starting with `docker-`.  If you want _way_ too much detail, the origin design seems to have been [here](https://github.com/docker/cli/issues/1534).
2. When a new Rancher Desktop is installed, the VM will be upgraded; this means that anything done on the `rancher-desktop` WSL distribution will get wiped out. This includes your docker server changes. (We keep container data in `rancher-desktop-data` instead for this reason.)  I'd recommend that if you want to use a different docker client, do that completely outside of Rancher Desktop, so that would at least not be affected by upgrades.

**Reply by ISimionEndava** [reply]
Cannot thank you enough. 

Where does the `docker` CLI reside after the Rancher Desktop install? I understand now that `docker-compose` will be placed in  `%USERPROFILE%\.docker\cli-plugins`, but where `docker` CLI would be placed?

**Reply by mook-as** [reply]
In PowerShell, try `Get-Command docker`. It should be in either:
- (Admin installs) `C:\Program Files\Rancher Desktop\resources\resources\win32\bin`
- (Per-user installs) `C:\Users\<>\AppData\Local\Programs\Rancher Desktop\resources\resources\win32\bin`

Note that that's off the top of my head, so you may need to adjust. Also, I recommend _not_ overwriting them, and just have a new one somewhere else and add it to your `%PATH%` _before_ the Rancher Desktop ones.

**Reply by ISimionEndava** [reply]
Thank you so much for every answer; you have been a real help. Thanks again for sharing your insights with me.
